### ISSUE TITLE
`$a:tt $($b:tt)*` is forbidden in macros

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#25436]
In an example like <http://is.gd/TKfdKf>:

```rust
macro_rules! foo {
  ( $a:tt $($b:tt)* ) => { }
}

fn main() { }
```

we report the following error:

```
<anon>:2:11: 2:19 error: `$a:tt` is followed by a sequence repetition, which is not allowed for `tt` fragments
<anon>:2   ( $a:tt $($b:tt)* ) => { }
                   ^~~~~~~~
error: aborting due to previous error
playpen: application terminated with error code 101
```

This is part of the macro future-proofing code but it is an unnecessary limitation. For fragments that represent at most one token tree (`tt`, `block`, `ident`, possibly others) there is no concern at all about being following by sequence repetition. This turns out to be a really big limitation in practice, since making macros that work over a generic sequence of token trees effectively allows for arbitrary parsing.

cc @pnkfelix

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by nikomatsakis** [comment]
I'd be happy to mentor this, if I don't fix it first.

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by nikomatsakis** [comment]
cc @cmr

**Comment by Stebalien** [comment]
I-ICE?

**Comment by bjadamson** [comment]
Curious why this was closed, looks like it's blocking lispy-rust.

**Comment by jdm** [comment]
@bjadamson It was fixed by #25444; do you still see this issue present in nightly?
